frame/assets/src/lib.rs

```diff
@@ -569,45 +589,207 @@ pub mod pallet {
 			Self::do_force_create(id, owner, is_sufficient, min_balance)
 		}
 
-		/// Destroy a class of fungible assets.
+		/// Start the process of destroying a class of fungible asset
+		/// start_destroy is the first in a series of extrinsics that should be called, to allow
+		/// destroying an asset.
 		///
 		/// The origin must conform to `ForceOrigin` or must be Signed and the sender must be the
 		/// owner of the asset `id`.
 		///
 		/// - `id`: The identifier of the asset to be destroyed. This must identify an existing
-		/// asset.
-		///
-		/// Emits `Destroyed` event when successful.
-		///
-		/// NOTE: It can be helpful to first freeze an asset before destroying it so that you
-		/// can provide accurate witness information and prevent users from manipulating state
-		/// in a way that can make it harder to destroy.
-		///
-		/// Weight: `O(c + p + a)` where:
-		/// - `c = (witness.accounts - witness.sufficients)`
-		/// - `s = witness.sufficients`
-		/// - `a = witness.approvals`
-		#[pallet::weight(T::WeightInfo::destroy(
-			witness.accounts.saturating_sub(witness.sufficients),
- 			witness.sufficients,
- 			witness.approvals,
- 		))]
-		pub fn destroy(
+		///   asset.
+		///
+		/// Assets must be freezed before calling start_destroy.
+		#[pallet::weight(T::WeightInfo::start_destroy())]
+		pub fn start_destroy(
+			origin: OriginFor<T>,
+			#[pallet::compact] id: T::AssetId,
+		) -> DispatchResult {
+			let maybe_check_owner = match T::ForceOrigin::try_origin(origin) {
+				Ok(_) => None,
+				Err(origin) => Some(ensure_signed(origin)?),
+			};
+			let _ = Asset::<T, I>::try_mutate_exists(
+				id,
+				|maybe_details| -> Result<(), DispatchError> {
+					let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+					if let Some(check_owner) = maybe_check_owner {
+						ensure!(details.owner == check_owner, Error::<T, I>::NoPermission);
+					}
+					ensure!(details.is_frozen, Error::<T, I>::BadWitness);
+					details.status = AssetStatus::Destroying;
+					// TODO: Remove previlleged roles. How?
+
+					Self::deposit_event(Event::Destroying { asset_id: id });
+					Ok(())
+				},
+			)?;
+			Ok(())
+		}
+
+		/// Destroy all accounts associated with a given asset.
+		/// `destroy_accounts` should only be called after `start_destroy` has been called, and the
+		/// asset is in a `Destroying` state
+		///
+		/// Due to weight restrictions, this function may need to be called multiple
+		/// times to fully destroy all accounts. It will destroy `RemoveKeysLimit` accounts at a
+		/// time.
+		///
+		/// The origin must conform to `ForceOrigin` or must be Signed and the sender must be the
+		/// owner of the asset `id`.
```

Comment:
At this point I would make it so that anyone can call the function. Since an asset can never come back from `Destroying`, once the `owner` puts it into that state there's no reason (AFAICT) to stop someone else from finishing the job.

Reply:
Should I also modify the extrinsic, such that it doesn't need an origin in the function signature [And also remove the origin check]?

Reply:
Every dispatchable function signature needs an origin, but you can do:
```rust
let _ = ensure_signed(origin)?;
```
For example, [Collectives's close](https://github.com/paritytech/substrate/blob/17c07af0b953b84dbe89341294e98e586f9b4591/frame/collective/src/lib.rs#L719).

Reply:
What do you think about @Chralt98 's comment?:
https://github.com/paritytech/substrate/pull/12310#pullrequestreview-1125106096 

> Shouldn't it be possible to destroy an asset from a hook like on_idle or on_initialize? We don't have an origin in that case. So maybe use pub functions like before do_destroy function and destroy as extrinsic. I think in that case, the user of pallet_assets doesn't need to have an origin available. That could be more flexible.